### PR TITLE
Adds DFS, Union-Find Array, Union-Find HashMap solutions (Sentence Similarity II)

### PR DESCRIPTION
**Pull request by altayhunoglu**
Resolves: https://github.com/spiralgo/algorithms/issues/66

When we represent the connections of the word pairs as a `graph`, it becomes easier to solve the problem using either DFS or Union-Find.

In the algorithm, we will mainly ask two questions to decide if the given sentences are similar:
1- Do the sentences have the same number of words?
2- Are the words (at the same index of the given sentences) similar?
    - If we use `DFS`, this means "Does the pair graph have a cycle for these two words?"
    - If we use `Union-Find`, this means "Are these two words in the same group with the same parent?"
    
    

> For example, sentence1= ["great", "acting", "skills"] and sentence2= ["fine", "drama", "talent"] are similar, if the similar word pairs are similar = [["great", "good"], ["fine", "good"], ["acting","drama"], ["skills","talent"]].

    
 
IMG-1635

### CODE REVIEW

**Comment by ErdemT09** [comment]
In the image, drawing the graph arrows undirected or double sided would have been more accurate, because if a is similar to b, then b is similar to a because of symmetry relation.

File: src/main/java/algorithms/curated170/medium/sentencesimilarity2/SentenceSimilarityIIDFS.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        if (words1.length != words2.length) {
+            return false;
+        }
+        graph = new HashMap();
+
+        for (List<String> pair : pairs) {
+            for (String p : pair) {
+                if (!graph.containsKey(p)) {
+                    graph.put(p, new ArrayList());
+                }
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/sentencesimilarity2/SentenceSimilarityIIDFS.java:24]: Small refactoring.
```suggestion
               graph.putIfAbsent(p, new ArrayList());
```

File: src/main/java/algorithms/curated170/medium/sentencesimilarity2/SentenceSimilarityIIUnionFindArray.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ 
+        Map<String, Integer> index = new HashMap();
+        int count = 0;
+        DSU dsu = new DSU(2 * pairs.size());
+        
+        for (List<String> pair : pairs) {
+            for (String p: pair){
+              if (!index.containsKey(p)) {
+                index.put(p, count++);
+            }
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/sentencesimilarity2/SentenceSimilarityIIUnionFindArray.java:22]: Again, putIfAbsent() method looks simpler.

File: src/main/java/algorithms/curated170/medium/sentencesimilarity2/SentenceSimilarityIIUnionFindMap.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        }
+        return find(parentWord);
+    }
+
+    void union(String w1, String w2) {
+        String parentWord1 = find(w1);
+        String parentWord2 = find(w2);
+ 
+        if (!parentWord1.equals(parentWord2)) {
+            parent.put(find(w1), find(w2));
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/sentencesimilarity2/SentenceSimilarityIIUnionFindMap.java:52]: Isn't this better in order not to repeat the process?
parent.put(parentWord2, parentWord2);

File: src/main/java/algorithms/curated170/medium/sentencesimilarity2/SentenceSimilarityIIUnionFindMap.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    String find(String w1) {
+        if(!parent.containsKey(w1)) return w1;
+        String parentWord = parent.get(w1);
+        if (parentWord.equals(w1)) {
+            return w1;
+        }
+        return find(parentWord);
+    }
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/sentencesimilarity2/SentenceSimilarityIIUnionFindMap.java:45]: We can perhaps apply parent compression here to reduce the repetiton of operations:
```suggestion
    String find(String w1) {
        if(!parent.containsKey(w1)) return w1;
        String parentWord = parent.get(w1);
        if (parentWord.equals(w1)) {
            return w1;
        }
        parent.put(w1, find(parenWord);
        return parent.get(w1);
    }
```

Comment (ErdemT09) [reply src/main/java/algorithms/curated170/medium/sentencesimilarity2/SentenceSimilarityIIUnionFindMap.java:45]:
```suggestion
    String find(String w1) {
        if(!parent.containsKey(w1)) return w1;
        String parentWord = parent.get(w1);
        if (!parentWord.equals(w1)) {
            parent.put(w1, find(parentWord));
        }
        return parent.get(w1);
    }
```

**Comment by altayhunoglu** [comment]
> In the image, drawing the graph arrows undirected or double sided would have been more accurate, because if a is similar to b, then b is similar to a because of symmetry relation.

The image has been updated with bidirectional arrows.
Thanks for the attention Erdem.

File: src/main/java/algorithms/curated170/medium/sentencesimilarity2/SentenceSimilarityIIUnionFindMap.java (hunk 3 above)
Comment (altayhunoglu) [reply src/main/java/algorithms/curated170/medium/sentencesimilarity2/SentenceSimilarityIIUnionFindMap.java]: Oke these are done.

**Comment by ErdemT09** [comment]
Basically a graph question perhaps covered as a string question. Thinking that way, questions like Longest Common Substring etc. are also graph questions to some degree. Anyway, approved.

**Review by ErdemT09 — APPROVED** [review]

**Pull request merged**